platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/AbstractLongList.java

```diff
@@ -441,25 +436,23 @@ protected final void writeHeader(final FileChannel fc) throws IOException {
      * @param subIndex   The sub index of the long in that chunk
      * @return The stored long value at given index
      */
-    protected abstract long lookupInChunk(final C chunk, final long subIndex);
+    protected abstract long lookupInChunk(@NonNull final C chunk, final long subIndex);
 
-    /**
-     * After invocation of this method, {@link AbstractLongList#get(long)}) calls
-     * will return {@link AbstractLongList#IMPERMISSIBLE_VALUE} for indices that
-     * are before {@code newMinValidIndex}.
-     * Also, a call to this method releases memory taken by unused byte buffers.
-     * Note that newMinValidIndex is allowed to exceed the current size of the list.
-     *
-     * @param newMinValidIndex minimal valid index of the list
-     */
+    /** {@inheritDoc} */
     @Override
-    public final void updateMinValidIndex(final long newMinValidIndex) {
+    public final void updateValidRange(final long newMinValidIndex, long maxValidIndex) {
         if (newMinValidIndex < 0) {
-            throw new IndexOutOfBoundsException("Min valid index " + newMinValidIndex + " must be non-negative");
+            throw new IndexOutOfBoundsException(MIN_VALID_INDEX_NON_NEGATIVE_MSG.formatted(newMinValidIndex));
+        }
+
+        if (maxValidIndex > maxLongs - 1) {
+            throw new IndexOutOfBoundsException(MAX_VALID_INDEX_LIMIT.formatted(maxValidIndex, maxLongs));
         }
 
         minValidIndex.set(newMinValidIndex);
-        shrinkIfNeeded(newMinValidIndex);
+        shrinkIfNeeded(newMinValidIndex, maxValidIndex);
+        // everything to the right of the maxValidIndex is going to be discarded, adjust the size accordingly
+        size.set(Math.min(maxValidIndex + 1, size.get()));
```

Comment:
It doesn't look good to me that we set the minimal index then remove chunks within the current `size` and only after that update `size` to the new limit.
Would it be cleaner and safer to update limits,  `minValidIndex` and `size`, and then call `shrinkIfNeeded(newMinValidIndex, maxValidIndex, oldSize)`?
Or maybe even inline `shrinkIfNeeded` as
```
        shrinkLeftSideIfNeeded(oldMinValidIndex, newMinValidIndex);
        shrinkRightSideIfNeeded(oldMaxValidIndex, newMaxValidIndex);
```

Reply:
Fair enough. I moved the update of `size` above the shrink calls and passed `oldMaxValidIndex` to `shrinkRightSideIfNeeded`. It turned out that `shrinkLeftSideIfNeeded` doesn't need `oldMinValidIndex`, so I decided not to introduce it.